### ISSUE TITLE
hdf5 general formatter does not work with lists

### ISSUE THREAD

**Issue by josenavas** [biocore/biom-format#638]
This line https://github.com/biocore/biom-format/blob/master/biom/table.py#L241 is passing as `shape=(len(md),)`. However, if each element in `md` is a list, it will fail with this error:

```
biom convert -i short.biom  -o shorthdf5.biom --to-hdf5
Traceback (most recent call last):
  File "/Users/jose/qiime_software/pyqi/scripts/pyqi", line 184, in <module>
    optparse_main(cmd_obj, argv[1:])
  File "/Users/jose/qiime_software/pyqi/pyqi/core/interfaces/optparse/__init__.py", line 275, in optparse_main
    result = optparse_cmd(local_argv[1:])
  File "/Users/jose/qiime_software/pyqi/pyqi/core/interface.py", line 41, in __call__
    return self._output_handler(cmd_result)
  File "/Users/jose/qiime_software/pyqi/pyqi/core/interfaces/optparse/__init__.py", line 250, in _output_handler
    opt_value)
  File "/Users/jose/qiime_software/biom-format/biom/interfaces/optparse/output_handler.py", line 80, in write_biom_table
    table.to_hdf5(f, generatedby())
  File "/Users/jose/qiime_software/biom-format/biom/table.py", line 3535, in to_hdf5
    self.group_metadata(axis='observation'), 'csr', compression)
  File "/Users/jose/qiime_software/biom-format/biom/table.py", line 3507, in axis_dump
    formatter[category](grp, category, md, compression)
  File "/Users/jose/qiime_software/biom-format/biom/table.py", line 243, in general_formatter
    compression=compression)
  File "/Users/jose/.virtualenvs/qiime/lib/python2.7/site-packages/h5py/_hl/group.py", line 94, in create_dataset
    dsid = dataset.make_new_dset(self, shape, dtype, data, **kwds)
  File "/Users/jose/.virtualenvs/qiime/lib/python2.7/site-packages/h5py/_hl/dataset.py", line 58, in make_new_dset
    raise ValueError("Shape tuple is incompatible with data")
ValueError: Shape tuple is incompatible with data
```

**Comment by ElDeveloper** [comment]
Just a heads up, this problem was also reported in the QIIME tracker see these example files and commands in the README to reproduce the problem:

https://www.dropbox.com/sh/4z67y4anz9dg1fb/AADoeoASMLqRq5D2YAGFe9UEa?dl=0

Original issue here: https://github.com/biocore/qiime/issues/2065

**Comment by ConstantinoSchillebeeckx** [comment]
Is there any update on this?

I updated multiple packages on my server a few days ago and now I'm running into this issue with my biom tables.

**Comment by wasade** [comment]
What type of metadata are associated with your samples or observations?

**Comment by ConstantinoSchillebeeckx** [comment]
Just taxonomy assignments & confidence: `Observation Metadata Categories: rdp_taxonomy; rdp_confidence`

I've also confirmed that my table is valid through `biom validate-table`

**Comment by ConstantinoSchillebeeckx** [comment]
I did some more digging on this issue. As part of our data pipeline, we keep the various biom table iterations as we filter/rarify/add metadata to it; I tested filtering samples on each of these iterations.

I am able to successfully run sample filtering on the version of the table that does not yet have OTU metadata added to it (done through `biom add-metadata`); however once we do add taxonomic metadata to the table, I'm no longer able to filter samples anymore.

For reference, the command being issued is `biom add-metadata -i otu_table_mc2_no_pynast_failures.biom -o otu_table_mc2_no_pynast_failures_rdp.biom --observation-metadata-fp rdp_taxa_qiime_format.txt --sc-separated rdp_taxonomy --float-fields rdp_confidence`

**Comment by wasade** [comment]
Sorry for the slow response -- the issue is that nested lists don't serialize easily into the HDF5 datasets. BIOM 1.0.0 is actually more permissive in this respect as JSON can handle it just fine. Our workaround was to specialcase the field `taxonomy` (and a few others) as we already knew it was a problem for them. In this case, the reason the problem is occurring is that `rdp_taxonomy` is not registered as a list of vlen str. 

There are a few workarounds. First, if `rdp_taxonomy` is changed to `taxonomy` then it *should* just work. The second is to issue calls directly to `Table.to_hdf5` and pass in custom formatters; this also entails passing in custom parsers on the call to `Table.from_hdf5`. 

Obviously this is not ideal. We've held off rewrites to the formatters and parsers as we're not sure what components may shift upstream to `scikit-bio`. But, at the same time, this is a development target which does need attention.

**Comment by ConstantinoSchillebeeckx** [comment]
Great, thanks for the work around. Are there any other `specialcase` fields I can use?

**Comment by wasade** [comment]
The current special case parsers are lined up [here](https://github.com/biocore/biom-format/blob/master/biom/table.py#L3596-L3600), and they have in kind formatters.

**Comment by wasade** [comment]
Closed by #737
